File: lib/nodejs/opentoken/src/attributes/person/AgeAttribute.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { BaseAttribute } from '../BaseAttribute';
import { RegexValidator } from '../validation/RegexValidator';
import { AgeRangeValidator } from '../validation/AgeRangeValidator';

/**
 * Represents the age attribute.
 *
 * This class extends BaseAttribute and provides functionality for working with
 * age fields. It recognizes "Age" as the primary alias for this attribute type.
 *
 * The attribute performs normalization on input values, trimming whitespace and
 * ensuring the value is a valid integer.
 *
 * The attribute also performs validation on input values, ensuring they:
 * - Are numeric (integers only)
 * - Fall within an acceptable range (0-120)
 */
export class AgeAttribute extends BaseAttribute {
  private static readonly NAME = 'Age';
  private static readonly ALIASES = ['Age'];
  private static readonly AGE_REGEX = /^\s*\d+\s*$/;

  constructor() {
    super(AgeAttribute.NAME, AgeAttribute.ALIASES, [
      new RegexValidator(AgeAttribute.AGE_REGEX),
      new AgeRangeValidator(),
    ]);
  }

  normalize(value: string): string {
    if (!value) {
      throw new Error('Age value cannot be null');
    }

    const trimmed = value.trim();

    try {
      const age = parseInt(trimmed, 10);
      if (isNaN(age)) {
        throw new Error(`Invalid age format: ${value}`);
      }
      return String(age);
    } catch (e) {
      throw new Error(`Invalid age format: ${value}`);
    }
  }
}

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/person/AgeAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.person;

import java.util.List;

import com.truveta.opentoken.attributes.general.IntegerAttribute;
import com.truveta.opentoken.attributes.validation.AgeRangeValidator;

/**
 * Represents the age attribute.
 * 
 * This class extends IntegerAttribute and provides functionality for working with
 * age fields. It recognizes "Age" as the primary alias for this attribute type.
 * 
 * The attribute performs normalization on input values, trimming whitespace and
 * ensuring the value is a valid integer (inherited from IntegerAttribute).
 * 
 * The attribute also performs validation on input values, ensuring they:
 * - Are numeric integers (inherited from IntegerAttribute)
 * - Fall within an acceptable range (0-120)
 */
public class AgeAttribute extends IntegerAttribute {

    private static final String NAME = "Age";
    private static final String[] ALIASES = new String[] { NAME };

    public AgeAttribute() {
        super(List.of(new AgeRangeValidator()));
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }

}
